ext/opcache/jit/zend_jit_ir.c

```diff
@@ -16685,19 +16763,125 @@ static bool zend_jit_may_be_in_reg(const zend_op_array *op_array, zend_ssa *ssa,
+		case 1: jit_ZVAL_COPY(jit, ZEND_ADDR_MEM_ZVAL(ZREG_RX, EX_NUM_TO_VAR(0)), MAY_BE_ANY & ~MAY_BE_REF, ZEND_ADDR_REF_ZVAL(op1_ref), op1_info, 1);
+	}
+
+	// Make call frame externally visible
+	ir_ref rx = jit_IP(jit);
+	ir_STORE(jit_CALL(rx, prev_execute_data), jit_FP(jit));
+	ir_STORE(jit_EG(current_execute_data), rx);
+
+	jit_observer_fcall_begin(jit, rx, observer_handler);
+
+	// JIT: fbc->internal_function.handler(new_frame, return_value)
+	ir_CALL_2(IR_VOID, ir_CONST_FC_FUNC(fbc->internal_function.handler), rx, res_ref);
+
+	jit_observer_fcall_end(jit, rx, res_ref);
+
+	ir_STORE(jit_EG(current_execute_data), ir_LOAD_A(jit_CALL(rx, prev_execute_data)));
+
+	for (int i = 0; i < call_num_args; ++i) {
+		if (zend_jit_needs_arg_dtor(fbc, i, NULL)) {
+			uint32_t offset = EX_NUM_TO_VAR(i);
+			zend_jit_addr var_addr = ZEND_ADDR_MEM_ZVAL(ZREG_RX, offset);
+
+			jit_ZVAL_PTR_DTOR(jit, var_addr, MAY_BE_ANY|MAY_BE_RC1|MAY_BE_RCN, 0, opline);
+		}
+	}
+
+	ir_ref allocated_path = IR_UNUSED;
+
+	// free the call frame
+	if (JIT_G(trigger) != ZEND_JIT_ON_HOT_TRACE ||
+	    !JIT_G(current_frame) ||
+	    !JIT_G(current_frame)->call ||
+	    !TRACE_FRAME_IS_NESTED(JIT_G(current_frame)->call)) {
+		// JIT: zend_vm_stack_free_call_frame(call);
+		// JIT: if (UNEXPECTED(ZEND_CALL_INFO(call) & ZEND_CALL_ALLOCATED))
+		ir_ref if_allocated = ir_IF(ir_AND_U8(
+			ir_LOAD_U8(ir_ADD_OFFSET(rx, offsetof(zend_execute_data, This.u1.type_info) + 2)),
+			ir_CONST_U8(ZEND_CALL_ALLOCATED >> 16)));
+		ir_IF_TRUE_cold(if_allocated);
+
+		ir_CALL_1(IR_VOID, ir_CONST_FC_FUNC(zend_jit_free_call_frame), rx);
+
+		allocated_path = ir_END();
+		ir_IF_FALSE(if_allocated);
+	}
+	ir_STORE(jit_EG(vm_stack_top), rx);
+
+	if (allocated_path) {
+		ir_MERGE_WITH(allocated_path);
+	}
+
+	if (reuse_ip) {
+		jit_STORE_IP(jit, ir_LOAD_A(jit_CALL(jit_FP(jit), call)));
+		if (jit->delayed_call_level) { // restore if saved
+			ir_STORE(jit_CALL(jit_FP(jit), call), ir_LOAD_A(jit_CALL(jit_IP(jit), prev_execute_data)));
+		}
+	} else {
+		// Note: conditional (if_unobserved) opline, zend_jit_set_last_valid_opline() may only be called if the opline is actually unconditionally updated
+		jit_STORE_IP(jit, ir_CONST_ADDR(jit->last_valid_opline ? jit->last_valid_opline : opline));
+	}
+
+	ir_ref skip = ir_END();
+	ir_IF_TRUE(if_unobserved);
+	return skip;
+}
+
+static void jit_frameless_icall0(zend_jit_ctx *jit, int checked_stack, const zend_op *opline)
 {
 	jit_SET_EX_OPLINE(jit, opline);
 
 	void *function = ZEND_FLF_HANDLER(opline);
 	zend_jit_addr res_addr = RES_ADDR();
 	ir_ref res_ref = jit_ZVAL_ADDR(jit, res_addr);
 	jit_set_Z_TYPE_INFO(jit, res_addr, IS_NULL);
+
+	ir_ref skip_observer = IR_UNUSED;
+	if (ZEND_OBSERVER_ENABLED) {
+		skip_observer = jit_frameless_observer(jit, checked_stack, opline, IR_UNUSED, 0, IR_UNUSED, 0, IR_UNUSED, 0, res_ref);
+	}
 	ir_CALL_1(IR_VOID, ir_CONST_ADDR((size_t)function), res_ref);
```

Comment:
@iluuu1994 do frameless functions always have the result?
I remember ICALL might use stack allocated zval for return value.

```
 cat x.php 
<?php
strpos($x, $y);
?>
$ sapi/cli/php -d opcache.opt_debug_level=0x30000  x.php

$_main:
     ; (lines=3, args=0, vars=2, tmps=1)
     ; (before optimizer)
     ; /home/dmitry/php/php-master/CGI-RELEASE-64/x.php:1-3
     ; return  [] RANGE[0..0]
0000 T2 = FRAMELESS_ICALL_2(strpos) CV0($x) CV1($y)
0001 FREE T2
0002 RETURN int(1)

$_main:
     ; (lines=2, args=0, vars=2, tmps=1)
     ; (after optimizer)
     ; /home/dmitry/php/php-master/CGI-RELEASE-64/x.php:1-3
0000 T2 = FRAMELESS_ICALL_2(strpos) CV0($x) CV1($y)
0001 RETURN int(1)
```

Is it OK to remove ``FREE``?